### PR TITLE
implemnt Prometheus sink

### PR DESCRIPTION
**Pull request by zirain**
this patch try to add native prometheus support instead of a statsd_exporter sidecar container.

@mattklein123 can you take a first look, I will finish the test if this's acceptable.

### CODE REVIEW

**Comment by mattklein123** [comment]
Sure sounds good. Please add docs also.

File: src/prometheusstats/prometheus_sink.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+			s.gauges[metricName] = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: metricName}, labelNames)
+		}
+
+		gauge := s.gauges[metricName]
+		gauge.WithLabelValues(labelValues...).Set(float64(value))
+	}
+}
+
+func (s *prometheusSink) FlushTimer(name string, value float64) {
+	m, labels, present := s.mapper.GetMapping(name, mapper.MetricTypeCounter)
```
Comment (SpecialYang) [inline src/prometheusstats/prometheus_sink.go:125]: Why is the type here not `MetricTypeTimer`？

Comment (zirain) [reply src/prometheusstats/prometheus_sink.go:125]: cpoy nit

**Comment by SpecialYang** [comment]
With the following rate limit policy, the process will crash.
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: ratelimit-config
data:
  config.yaml: |
    domain: ratelimit
    descriptors:
      - key: PATH
        value: "/version"
        rate_limit:
          unit: minute
          requests_per_unit: 10
        descriptors:
        - key: test
          value: true
          rate_limit:
            unit: minute
            requests_per_unit: 2
``` 

```
panic: inconsistent label cardinality: expected 3 label values but got 2 in []string{"ratelimit", "PATH_/version"}

goroutine 21 [running]:
github.com/prometheus/client_golang/prometheus.(*CounterVec).WithLabelValues(...)
	/go/pkg/mod/github.com/prometheus/client_golang@v1.19.1/prometheus/counter.go:284
github.com/envoyproxy/ratelimit/src/prometheusstats.(*prometheusSink).FlushCounter(0xc00013c4b0, {0xc0004c20c0?, 0xc0004ce690?}, 0x1b)
	/ratelimit/src/prometheusstats/prometheus_sink.go:122 +0x4ec
github.com/lyft/gostats.(*statStore).Flush.func1({0xd434e0?, 0xc0004dd580?}, {0xe36580?, 0xc000392d00?})
	/go/pkg/mod/github.com/lyft/gostats@v0.4.14/stats.go:386 +0x76
sync.(*Map).Range(0xc0000d36e0?, 0xc0004a7f08)
	/usr/local/go/src/sync/map.go:477 +0x1f8
github.com/lyft/gostats.(*statStore).Flush(0xc0004d6320)
	/go/pkg/mod/github.com/lyft/gostats@v0.4.14/stats.go:383 +0xbf
github.com/lyft/gostats.(*statStore).StartContext(0xc0004d6320, {0x1083980, 0x1888980}, 0xc00013d770)
	/go/pkg/mod/github.com/lyft/gostats@v0.4.14/stats.go:367 +0x35
github.com/lyft/gostats.(*statStore).Start(0x0?, 0x0?)
	/go/pkg/mod/github.com/lyft/gostats@v0.4.14/stats.go:373 +0x25
created by github.com/envoyproxy/ratelimit/src/service_cmd/runner.NewRunner in goroutine 1
	/ratelimit/src/service_cmd/runner/runner.go:77 +0x74b
```

**Comment by zirain** [comment]
> With the following rate limit policy, the process will crash.
> 
> ```
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: ratelimit-config
> data:
>   config.yaml: |
>     domain: ratelimit
>     descriptors:
>       - key: PATH
>         value: "/version"
>         rate_limit:
>           unit: minute
>           requests_per_unit: 10
>         descriptors:
>         - key: test
>           value: true
>           rate_limit:
>             unit: minute
>             requests_per_unit: 2
> ```
> 
> ```
> panic: inconsistent label cardinality: expected 3 label values but got 2 in []string{"ratelimit", "PATH_/version"}
> 
> goroutine 21 [running]:
> github.com/prometheus/client_golang/prometheus.(*CounterVec).WithLabelValues(...)
> 	/go/pkg/mod/github.com/prometheus/client_golang@v1.19.1/prometheus/counter.go:284
> github.com/envoyproxy/ratelimit/src/prometheusstats.(*prometheusSink).FlushCounter(0xc00013c4b0, {0xc0004c20c0?, 0xc0004ce690?}, 0x1b)
> 	/ratelimit/src/prometheusstats/prometheus_sink.go:122 +0x4ec
> github.com/lyft/gostats.(*statStore).Flush.func1({0xd434e0?, 0xc0004dd580?}, {0xe36580?, 0xc000392d00?})
> 	/go/pkg/mod/github.com/lyft/gostats@v0.4.14/stats.go:386 +0x76
> sync.(*Map).Range(0xc0000d36e0?, 0xc0004a7f08)
> 	/usr/local/go/src/sync/map.go:477 +0x1f8
> github.com/lyft/gostats.(*statStore).Flush(0xc0004d6320)
> 	/go/pkg/mod/github.com/lyft/gostats@v0.4.14/stats.go:383 +0xbf
> github.com/lyft/gostats.(*statStore).StartContext(0xc0004d6320, {0x1083980, 0x1888980}, 0xc00013d770)
> 	/go/pkg/mod/github.com/lyft/gostats@v0.4.14/stats.go:367 +0x35
> github.com/lyft/gostats.(*statStore).Start(0x0?, 0x0?)
> 	/go/pkg/mod/github.com/lyft/gostats@v0.4.14/stats.go:373 +0x25
> created by github.com/envoyproxy/ratelimit/src/service_cmd/runner.NewRunner in goroutine 1
> 	/ratelimit/src/service_cmd/runner/runner.go:77 +0x74b
> ```

can you share more logs?

**Comment by zirain** [comment]
@SpecialYang can you share the reproduce steps?

**Comment by zirain** [comment]
There are a lot of problems with this implementation, I'd like close this one first, and proposal a new one later.

**Pull request closed without merging**